### DISCUSSION TITLE
Does package-based invalidation work based on version number or hash?

### DISCUSSION THREAD

**Discussion by joelnitta** [ropensci/targets#926] [Help] [answered]
For [package-based invalidation](https://books.ropensci.org/targets/packages.html#package-based-invalidation), it is a little unclear to me if the invalidation only happens when there is a new version of the package (change to `Version` of `DESCRIPTION`) or if it hash-based, like the rest of {targets} invalidation.

Specifically, I am trying to use package-based invalidation in a project where I install [my own in-development package from github](https://github.com/fernphy/pteridocat). I often push changes to the package without bumping the version. But when I re-install the package (with {renv}) and run `tar_make()` again, it does not seem to recognize the package as invalidated. In particular, I observe this when loading a data object from the package:

```
tar_plan(
  # Load data ----
  # Pteridocat taxonomic database
  pteridocat = pteridocat::pteridocat,
...
```

**Comment by wlandau** [comment, marked as answer]
Package-based invalidation through `tar_option_set(imports = "...")` acts by hashing each of the objects and functions in the namespace of the package. So if your update to `pteridocat` did not change the hash of `pteridocat::pteridocat`, then it would not cause invalidation. The dependency graph shows all the objects tracked this way.

**Reply by joelnitta** [reply]
Hm, that's what I thought. The strange thing is, my update did change the hash of `pteridocat`, but it didn't get invalidated. I had to run `tar_invalidate(pteridocat)` manually, then the changes got picked up from there.

**Reply by wlandau** [reply]
Happy to look at a reprex.

**Reply by joelnitta** [reply]
Sure thing.

Code to generate reprex:

```r
targets::tar_reprex(
  pipeline = {
    tar_option_set(imports = "pteridocat")
    list(
      tar_target(pteridocat, pteridocat::pteridocat)
    )
  },
  run = {
    # install package at most recent commit
    renv::install("fernphy/pteridocat@16e2ae1ae902621a3adfece306a3965f391bb722")
    tar_make()
    # check hash of pteridocat in targets cache
    digest::digest(tar_read(pteridocat))
    # install pacakge at earlier commit (pteridocat content is different)
    renv::install("fernphy/pteridocat@7f7a4a9bf3172b38138333e9973c1c308737e88a")
    tar_make()
    digest::digest(tar_read(pteridocat))
    # manually invalidate and run again
    tar_invalidate(pteridocat)
    tar_make()
    # check hash of pteridocat in targets cache
    digest::digest(tar_read(pteridocat))
  }
)
```

Reprex:

``` r
library(targets)
tar_script({
    tar_option_set(imports = "pteridocat")
    list(tar_target(pteridocat, pteridocat::pteridocat))
})
renv::install("fernphy/pteridocat@16e2ae1ae902621a3adfece306a3965f391bb722")
#> Installing pteridocat [1.0.0] ...
#>  OK [linked cache]
tar_make()
#> • start target pteridocat
#> • built target pteridocat
#> • end pipeline: 0.701 seconds
digest::digest(tar_read(pteridocat))
#> [1] "902fa0d16ed11561b9d65681734675f6"
renv::install("fernphy/pteridocat@7f7a4a9bf3172b38138333e9973c1c308737e88a")
#> Installing pteridocat [1.0.0] ...
#>  OK [linked cache]
tar_make()
#> ✔ skip target pteridocat
#> ✔ skip pipeline: 0.042 seconds
digest::digest(tar_read(pteridocat))
#> [1] "902fa0d16ed11561b9d65681734675f6"
tar_invalidate(pteridocat)
tar_make()
#> • start target pteridocat
#> • built target pteridocat
#> • end pipeline: 0.7 seconds
digest::digest(tar_read(pteridocat))
#> [1] "4bbb24a2464726cc7efdafe5bc7708af"
```

<sup>Created on 2022-09-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details>
<summary>sessionInfo()</summary>

```
R version 4.2.0 (2022-04-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Monterey 12.5.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
[1] targets_0.12.1

loaded via a namespace (and not attached):
 [1] styler_1.7.0         tidyselect_1.1.2     xfun_0.32            remotes_2.4.2        rematch2_2.1.2       purrr_0.3.4          cranlike_1.0.2      
 [8] vctrs_0.4.1          htmltools_0.5.2      crancache_0.0.0.9001 yaml_2.3.5           utf8_1.2.2           blob_1.2.3           rlang_1.0.5         
[15] pkgbuild_1.3.1       R.oo_1.25.0          pillar_1.8.1         R.utils_2.12.0       glue_1.6.2           withr_2.5.0          DBI_1.1.3           
[22] rappdirs_0.3.3       bit64_4.0.5          lifecycle_1.0.1      R.cache_0.15.0       R.methodsS3_1.8.2    evaluate_0.16        codetools_0.2-18    
[29] memoise_2.0.1        knitr_1.40           callr_3.7.2          fastmap_1.1.0        ps_1.7.1             curl_4.3.2           fansi_1.0.3         
[36] highr_0.9            Rcpp_1.0.9           clipr_0.8.0          renv_0.15.5          backports_1.4.1      cachem_1.0.6         desc_1.4.1          
[43] debugme_1.1.0        fs_1.5.2             bit_4.0.4            digest_0.6.29        processx_3.7.0       parsedate_1.3.0      rprojroot_2.0.3     
[50] cli_3.3.0            tools_4.2.0          magrittr_2.0.3       base64url_1.4        tibble_3.1.8         RSQLite_2.2.16       crayon_1.5.1        
[57] pkgconfig_2.0.3      data.table_1.14.2    prettyunits_1.1.1    reprex_2.0.1         rmarkdown_2.14       rstudioapi_0.14      R6_2.5.1            
[64] pteridocat_1.0.0     igraph_1.3.4         compiler_4.2.0      
```

</details>

**Reply by wlandau** [reply]
Thanks, now I think I see what is happening. First, `targets` uses `getNamespace()` to identify the objects in an "imports" package:

https://github.com/ropensci/targets/blob/c8d934ca21c718851dc3e1f7f21498eae5ab97ab/R/class_imports.R#L24

Unfortunately - and I didn't realize this - [datasets are not necessarily in the namespace](https://stat.ethz.ch/pipermail/r-devel/2012-August/064719.html):

```r
ls(getNamespace("pteridocat"), all.names=TRUE)
> [1] ".__NAMESPACE__."      ".__S3MethodsTable__." ".packageName" 
```

I thought about `as.environment()` as an alternative, but it only lists exported objects, and it only works if the package is attached.


```r
ls(as.environment("package:pteridocat"), all.names = TRUE)
#> Error in as.environment(pos) : 
#>   no item called "as.environment("package:pteridocat")" on the search list
#> In addition: Warning message:
#> In ls(as.environment("package:pteridocat"), all.names = TRUE) :
#>   ‘as.environment("package:pteridocat")’ converted to character string
library(pteridocat)
ls(as.environment("package:pteridocat"), all.names = TRUE)
#> [1] "pteridocat"
```

As of 98880e71c06e9e7778a437293780221333dd6af4, `targets` now imports datasets from packages. One caveat is that namespaced calls, e.g. `pteridocat::pteridocat`, will not have edges in the dependency graph. Objects and datasets are imported as if they are part of the global environment, so it is necessary to write e.g. `tar_target(data_name, pteridocat)`. This is because of the limitations of dependency detection via static code analysis based on `codetools::findGlobals()`, which treats `::` and `:::` as functions and not part of the names of symbols. To import selected objects directly, it is best to make them global objects in `_targets.R` and refer to those.

```r
# _targets.R file
global_dataset <- pteridocat::pteridocat
list(
  tar_target(data_target_name, global_dataset)
)
```

**Reply by joelnitta** [reply]
Works for me, thanks!
